File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/group/Modified.java
package de.cubbossa.pathfinder.group;

import de.cubbossa.pathfinder.misc.NamespacedKey;
import java.util.Collection;
import java.util.Optional;

public interface Modified {

  Collection<Modifier> getModifiers();

  <M extends Modifier> boolean hasModifier(Class<M> modifierClass);

  <M extends Modifier> boolean hasModifier(NamespacedKey modifierType);

  default <M extends Modifier> void addModifier(M modifier) {
    addModifier(modifier.getKey(), modifier);
  }

  void addModifier(NamespacedKey key, Modifier modifier);

  <M extends Modifier> Optional<M> getModifier(NamespacedKey key);

  <M extends Modifier> void removeModifier(Class<M> modifierClass);

  <M extends Modifier> void removeModifier(M modifier);

  <M extends Modifier> void removeModifier(NamespacedKey key);

  void clearModifiers();
}

File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/group/Modified.kt
package de.cubbossa.pathfinder.group

import de.cubbossa.pathfinder.misc.NamespacedKey

interface Modified {
    val modifiers: Collection<Modifier>

    fun <M : Modifier> hasModifier(modifierClass: Class<M>): Boolean

    fun <M : Modifier> hasModifier(modifierType: NamespacedKey): Boolean

    fun <M : Modifier> addModifier(modifier: M) {
        addModifier(modifier.key, modifier)
    }

    fun addModifier(key: NamespacedKey, modifier: Modifier)

    fun <M : Modifier> getModifier(key: NamespacedKey): M?

    fun <M : Modifier> removeModifier(modifierClass: Class<M>)

    fun <M : Modifier> removeModifier(modifier: M)

    fun removeModifier(key: NamespacedKey)

    fun clearModifiers()
}

inline fun <reified M : Modifier> Modified.hasModifier(): Boolean {
    return hasModifier(M::class.java)
}
